Documentation Type Lookup Service tutorial [17108] (#17)
eProsima/DDS-Recorder: docs/rst/recording/tutorials/basic_example.rst

.. _tutorials_basic_example:

#################################################

> Recording DynamicType data from a DDS application

Reviewer: Configuring Fast DDS DynamicTypes for data recording

#################################################

.. contents::
    :local:
    :backlinks: none
    :depth: 2

**********
Background
**********

> Since the DDS Recorder only supports dynamic types, DDS Publisher using DynamicTypes for the data type definition.

Reviewer: ```suggestion
Currently, the DDS Recorder only stores DDS data whose data type is set as a DynamicType. The reason for this is that, without the need for the user to set the data type in the DDS Recorder, the DDS Recorder can access it via the type lookup service.  
Therefore, user's application DDS Publisher must be configured to use DynamicTypes for the data type definition.
```

Thus, the user will be able to record the published data using the |ddsrecorder| tool of the |eddsrecord| software.

> Moreover, the example implements a DDS Subscriber that will receive any kind of data published by the Publisher.
> For that task, the DDS Publisher and the DDS Subscriber implement a TypeLookupService that send the TypeInformation of the data types used by the publisher.

Reviewer: ```suggestion
```

Author: What's the difference?

Reviewer: Just remove it

The source code of this tutorial is provided `here <https://github.com/eProsima/DDS-Recorder/tree/main/resources/dds/TypeLookupService>`_ with an explanation of how to build and run it.

This tutorial focuses on how to send the data type information using Fast DDS DynamicTypes and other relevant aspects of DynamicTypes.

> For more information about how to create the workspace with a basic DDS Publisher and a basic DDS Subscriber, please refer to `Writing a simple C++ publisher and subscriber application <https://fast-dds.docs.eprosima.com/en/latest/fastdds/getting_started/simple_app/simple_app.html>`_ .

Reviewer: ```suggestion
More specifically, this tutorial implements a DDS Publisher configured to send its data type, a DDS Subscriber that collects the data type and is able to read the incoming data, and a DDS Recorder is launched to save all the data published on the network. 
For more information about how to create the workspace with a basic DDS Publisher and a basic DDS Subscriber, please refer to `Writing a simple C++ publisher and subscriber application <https://fast-dds.docs.eprosima.com/en/latest/fastdds/getting_started/simple_app/simple_app.html>`_ .
```

If |eddsrecord| was installed using the `recommended installation <https://dds-recorder.readthedocs.io/en/latest/rst/installation/docker.html>`_ the environment is source by default, otherwise, just remember to source it in every terminal in this tutorial:

.. code-block:: bash

    source <path-to-fastdds-installation>/install/setup.bash
    source <path-to-ddspipe-installation>/install/setup.bash
    source <path-to-ddsrecorder-installation>/install/setup.bash

*********************
Generating data types
*********************

`eProsima Fast DDS-Gen <https://fast-dds.docs.eprosima.com/en/latest/fastddsgen/introduction/introduction.html>`_ is a Java application that generates *eProsima Fast DDS* source code using the data types defined in an IDL (Interface Definition Language) file.
When generating the Types using *eProsima Fast DDS Gen*, the option ``-typeobject`` must be added in order to generate the needed code to fill the ``TypeInformation`` data.

The expected argument list of the application is:

.. code-block:: bash

    fastddsgen -typeobject MyType.idl

*************
DDS Publisher
*************

> The DDS publisher is acting as a server of types.

Reviewer: ```suggestion
The DDS publisher will be configured to act as a server of the data types of the data it publishes.
Thus, any application ready to read this data types will be able to deserialize the incoming data using the type information received. 
```

> Fast DDS does not send the DataType information by default, it must be configured to do so.

Reviewer: ```suggestion
However, *Fast DDS* does not send the data type information by default, it must be configured to do so.
```

Complex types
============

For complex types, it is required to use ``TypeInformation`` mechanism. In the *eProsima Fast DDS* ``DomainParticipant`` set the following QoS in order to send this information:

.. code-block:: bash

    DomainParticipantQos pqos;
    pqos.wire_protocol().builtin.typelookup_config.use_server = true;

Native types
============

> For native types *eProsima Fast DDS* will send the ``TypeObject`` by default.

Reviewer: Remove as it is already explained it in the examining the code section

Data types
==========

At the moment, there are two data types that can be used:

* `HelloWorld.idl <https://github.com/eProsima/DDS-Recorder/blob/main/resources/dds/TypeLookupService/types/hello_world/HelloWorld.idl>`_

.. literalinclude:: ../../../../resources/dds/TypeLookupService/types/hello_world/HelloWorld.idl

* `Complete.idl <https://github.com/eProsima/DDS-Recorder/blob/main/resources/dds/TypeLookupService/types/complete/Complete.idl>`_

.. literalinclude:: ../../../../resources/dds/TypeLookupService/types/complete/Complete.idl

Examining the code
==================

This section explains the C++ source code of the DDS Publisher, which can also be found `here <https://github.com/eProsima/DDS-Recorder/blob/main/resources/dds/TypeLookupService>`_.

> For simplicity is going to be explain just the code related to the ``HelloWorld`` type.
> 
> The private data members of the class defines the DDS Topic, DataTypeKind, DDS Topic type and DynamicType.

Reviewer: ```suggestion
The private data members of the class defines the DDS Topic, ``DataTypeKind``, DDS Topic type and DynamicType.
The ``DataTypeKind`` defines the type to be used by the application (``HelloWorld`` or ``Complete``).
For simplicity, this tutorial only covers the code related to the ``HelloWorld`` type.
```

.. literalinclude:: ../../../../resources/dds/TypeLookupService/TypeLookupServicePublisher.h
    :language: C++
    :lines: 139-146

The next lines show the constructor of the ``TypeLookupServicePublisher`` class that implements the publisher.
The publisher is created with the topic and data type to use.

.. literalinclude:: ../../../../resources/dds/TypeLookupService/TypeLookupServicePublisher.cpp
    :language: C++
    :lines: 45-54

Inside the ``TypeLookupServicePublisher`` constructor are defined the DomainParticipantQos.

> As the publisher act as a server of types, set the QoS in order to send this information.

Reviewer: ```suggestion
As the publisher act as a server of types, its QoS must be configured to send this information.
```

Set ``use_client`` to ``false`` and ``use_server`` to ``true``.

.. literalinclude:: ../../../../resources/dds/TypeLookupService/TypeLookupServicePublisher.cpp
    :language: C++
    :lines: 58-62

Next, we register the type in the participant:

1. Generate the dynamic type through ``generate_helloworld_type_()`` explained below.
2. Set the data type.
3. Create the ``TypeSupport`` with the dynamic type previously created.

> 4. Set the ``type`` to ``TypeInformation`` and not to ``TypeObject`` since we want to send the information of the type and not the object.

Reviewer: ```suggestion
4. Configure the ``type`` to fill automatically the ``TypeInformation`` and not ``TypeObject`` to be compliant with `DDS-XTypes 1.2. <https://www.omg.org/spec/DDS-XTypes/1.2>`_ standard.
```

> * This answer `DDS-XTypes 1.2. <https://www.omg.org/spec/DDS-XTypes/1.2>`_

Reviewer: ```suggestion
```

.. literalinclude:: ../../../../resources/dds/TypeLookupService/TypeLookupServicePublisher.cpp
    :language: C++
    :lines: 73-95

> The function ``generate_helloworld_type_()`` returns the dynamic type generated with the object and id of the type.

Reviewer: ```suggestion
The function ``generate_helloworld_type_()`` returns the dynamic type generated with the ``TypeObject`` and ``TypeIdentifier`` of the type.
```

.. literalinclude:: ../../../../resources/dds/TypeLookupService/TypeLookupServicePublisher.cpp
    :language: C++
    :lines: 256-271

> Then are initialized the Publisher, DDS Topic and DDS DataWriter.

Reviewer: ```suggestion
Then we initialized the Publisher, DDS Topic and DDS DataWriter.
```

To make the publication, the public member function ``publish()`` is implemented:

> 1. It create the variable that will contain the user data, ``dynamic_data_``.

Reviewer: ```suggestion
1. It creates the variable that will contain the user data, ``dynamic_data_``.
```

2. Fill that variable with the function ``fill_helloworld_data_(msg)``, explained below.

.. literalinclude:: ../../../../resources/dds/TypeLookupService/TypeLookupServicePublisher.cpp
    :language: C++
    :lines: 229-254

The function ``fill_helloworld_data_()`` returns the data to be sent with the information filled in.

> First, it is created the ``Dynamic_ptr`` that will be filled in and return.

Reviewer: ```suggestion
First, the ``Dynamic_ptr`` that will be filled in and returned is created.
```

> With the use of ``DynamicDataFactory`` we create the data that corresponds to our data type.

Reviewer: ```suggestion
Using the ``DynamicDataFactory`` we create the data that corresponds to our data type.
```

Finally, data variables are assigned, in this case, ``index`` and ``message``.

.. literalinclude:: ../../../../resources/dds/TypeLookupService/TypeLookupServicePublisher.cpp
    :language: C++
    :lines: 293-307

**************
DDS Subscriber
**************

> The DDS Subscriber is acting as a client of types.
> Thus, the subscriber does not need to know the type.

Reviewer: ```suggestion
The DDS Subscriber is acting as a client of types, i.e. the subscriber will not know the types beforehand and it will discovery the data type via the type lookup service implemented on the publisher side. 
```

Examining the code
==================

> This section explains the C++ source code of the DDS Publisher, which can also be found `here <https://github.com/eProsima/DDS-Recorder/blob/main/resources/dds/TypeLookupService>`_.

Reviewer: ```suggestion
This section explains the C++ source code of the DDS Subscriber, which can also be found `here <https://github.com/eProsima/DDS-Recorder/blob/main/resources/dds/TypeLookupService>`_.
```

The private data members of the class defines the DDS Topic, DDS Topic type and DynamicType.

.. literalinclude:: ../../../../resources/dds/TypeLookupService/TypeLookupServiceSubscriber.h
    :language: C++
    :lines: 105-110

The next lines show the constructor of the ``TypeLookupServiceSubscriber`` class that implements the subscriber setting the topic name as the one configured in the
publisher side.

.. literalinclude:: ../../../../resources/dds/TypeLookupService/TypeLookupServiceSubscriber.cpp
    :language: C++
    :lines: 45-53

> Inside the ``TypeLookupServiceSubscriber`` constructor are defined the DomainParticipantQos.

Reviewer: ```suggestion
The ``DomainParticipantQos`` are defined inside the ``TypeLookupServiceSubscriber`` constructor.
```

> As the publisher act as a client of types, set the QoS in order to receive this information.

Reviewer: ```suggestion
As the subscriber act as a client of types, set the QoS in order to receive this information.
```

Set ``use_client`` to ``true`` and ``use_server`` to ``false``.

.. literalinclude:: ../../../../resources/dds/TypeLookupService/TypeLookupServiceSubscriber.cpp
    :language: C++
    :lines: 57-61

> Then is initialized the Subscriber.

Reviewer: ```suggestion
Then, the Subscriber is initialized.
```

> Inside ``on_data_available()`` it is created the ``DynamicData_ptr`` where the samples received will be read.

Reviewer: ```suggestion
Inside ``on_data_available()`` callback function the ``DynamicData_ptr`` is created, which will be filled with the actual data received. 
```

> As in the publisher, it is used ``DynamicDataFactory`` for the creation of the data that corresponds to our data type.

Reviewer: ```suggestion
As in the publisher, the``DynamicDataFactory`` is used for the creation of the data that corresponds to our data type.
```

.. literalinclude:: ../../../../resources/dds/TypeLookupService/TypeLookupServiceSubscriber.cpp
    :language: C++
    :lines: 143-170

> The function ``on_type_information_received()`` detect the topics received and discover it if they have the same as the topic that we are waiting for.

Reviewer: ```suggestion
The function ``on_type_information_received()`` detects if new topic information has been received in order to proceed to register the topic in case it has the same name as the expected one. 
```

> Discover the topic means set the topic as discover and register it using the function ``register_remote_type_callback_()`` explained below.

Reviewer: ```suggestion
To register a remote topic, function ``register_remote_type_callback_()`` is used.
```

> Once the topic has been discovered and registered, it is created a DataReader on this topic.

Reviewer: ```suggestion
Once the topic has been discovered and registered, we create a DataReader on this topic.
```

.. literalinclude:: ../../../../resources/dds/TypeLookupService/TypeLookupServiceSubscriber.cpp
    :language: C++
    :lines: 172-212

> Finally, the function ``register_remote_type_callback_()`` is in charge of register the topic received.

Reviewer: ```suggestion
The function ``register_remote_type_callback_()``, which is in charge of register the topic received, is explained below. 
```

> First of all, it creates a ``TypeSupport`` with the corresponding type and register it into the participant.

Reviewer: ```suggestion
First, it creates a ``TypeSupport`` with the corresponding type and registers it into the participant.
```

> Then, it creates the DDS Topic with the topic name set in the creation of the Subscriber and the topic type previously register.

Reviewer: ```suggestion
Then, it creates the DDS Topic with the topic name set in the creation of the Subscriber and the topic type previously registered.
```

Finally, it creates the DataReader of that topic.

.. literalinclude:: ../../../../resources/dds/TypeLookupService/TypeLookupServiceSubscriber.cpp
    :language: C++
    :lines: 278-304

***********************
Running the application
***********************

Recording samples with DDS Recorder
===================================

Open two terminals:

> In the first terminal, run:

Reviewer: ```suggestion
* In the first terminal, run an instance of the described publisher. 
```

> .. code-block:: bash

Reviewer: Notice that I suggested to add a list here so be careful with indentations

    source install/setup.bash
    cd DDS-Recorder/build/TypeLookupService
    ./TypeLookupService

In the second terminal, run the ddsrecorder:

.. code-block:: bash

    source install/setup.bash

>     ddsrecorder

Reviewer: ```suggestion
    ddsrecorder
    
Stop the |ddsrecorder| at any time to save the output MCAP file containing the recorded data. 
```

Publisher <-> Subscriber
========================

Open two terminals:

> In the first terminal, run the DDS Publisher:

Reviewer: Same here, add a list item for each terminal

.. code-block:: bash

    source install/setup.bash
    cd DDS-Recorder/build/TypeLookupService
    ./TypeLookupService --entity publisher

In the second terminal, run the DDS Subscriber:

.. code-block:: bash

    source install/setup.bash
    cd DDS-Recorder/build/TypeLookupService
    ./TypeLookupService --entity subscriber

> .. figure:: /resources/tutorials/basic_publisher_subscriber.png

Reviewer: ```suggestion
At this point, we observe that the data published reach the subscriber and it can access to the content of the sample received. 

.. figure:: /resources/tutorials/basic_publisher_subscriber.png
```
